huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go

```diff
@@ -0,0 +1,293 @@
+	}
+
+	listListenersRespJson, err := json.Marshal(listListenersResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	var listListenersRespBody interface{}
+	err = json.Unmarshal(listListenersRespJson, &listListenersRespBody)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	uuid, err := uuid.GenerateUUID()
+	if err != nil {
+		return diag.Errorf("unable to generate ID: %s", err)
+	}
+	d.SetId(uuid)
+
+	var mErr *multierror.Error
+	mErr = multierror.Append(
+		mErr,
+		d.Set("listeners", filterListListenersResponseBody(flattenListListenersResponseBody(listListenersRespBody), d)),
+	)
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func flattenListListenersResponseBody(resp interface{}) []interface{} {
+	if resp == nil {
+		return nil
+	}
+
+	curJson := utils.PathSearch("listeners", resp, make([]interface{}, 0))
+	curArray := curJson.([]interface{})
+	rst := make([]interface{}, 0, len(curArray))
+	for _, v := range curArray {
+		rst = append(rst, map[string]interface{}{
+			"id":              utils.PathSearch("id", v, nil),
+			"name":            utils.PathSearch("name", v, nil),
+			"status":          utils.PathSearch("status", v, nil),
+			"description":     utils.PathSearch("description", v, nil),
+			"protocol":        utils.PathSearch("protocol", v, nil),
+			"port_ranges":     flattenPortRanges(utils.PathSearch("port_ranges", v, make([]interface{}, 0))),
+			"client_affinity": utils.PathSearch("client_affinity", v, nil),
+			"accelerator_id":  utils.PathSearch("accelerator_id", v, nil),
+			"tags":            utils.FlattenTagsToMap(utils.PathSearch("tags", v, nil)),
+			"created_at":      utils.PathSearch("created_at", v, nil),
+			"updated_at":      utils.PathSearch("updated_at", v, nil),
+		})
+	}
+	return rst
+}
+
+func flattenPortRanges(raw interface{}) []map[string]interface{} {
+	curArray := raw.([]interface{})
+	result := make([]map[string]interface{}, len(curArray))
+	for i, portRanges := range curArray {
+		result[i] = map[string]interface{}{
+			"from_port": utils.PathSearch("from_port", portRanges, nil),
+			"to_port":   utils.PathSearch("to_port", portRanges, nil),
+		}
+	}
+	return result
+}
+
+func filterListListenersResponseBody(all []interface{}, d *schema.ResourceData) []interface{} {
+	rst := make([]interface{}, 0, len(all))
+	for _, v := range all {
+		if param, ok := d.GetOk("protocol"); ok &&
+			fmt.Sprint(param) != utils.PathSearch("protocol", v, nil) {
+			continue
+		}
+
+		portRanges := utils.PathSearch("port_ranges", v, nil).([]map[string]interface{})
+		if param, ok := d.GetOk("from_port"); ok && len(filterByPortRange(portRanges, param.(int), "from_port")) == 0 {
+			continue
+		}
+
+		if param, ok := d.GetOk("to_port"); ok && len(filterByPortRange(portRanges, param.(int), "to_port")) == 0 {
+			continue
+		}
+
+		rst = append(rst, v)
+	}
+	return rst
+}
+
+func filterByPortRange(portRanges []map[string]interface{}, target int, port string) []map[string]interface{} {
```

Comment:
I think the return value of this method only needs to be a boolean value.

Reply:
done